File: packages/csharp/src/GofPatterns/Structural.cs
namespace GofPatterns;

/// <summary>
/// Composite — treat a tree of nodes like a single node.
/// <para><see cref="Walk"/> is lazy, so LINQ and an early <c>First</c> stop the traversal.</para>
/// </summary>
public sealed class Composite<T>(T value, IEnumerable<Composite<T>>? children = null)
{
    private readonly List<Composite<T>> children = children?.ToList() ?? [];

    public T Value { get; } = value;

    public IReadOnlyList<Composite<T>> Children => this.children;

    public Composite<T> Add(params Composite<T>[] nodes)
    {
        this.children.AddRange(nodes);
        return this;
    }

    public bool Remove(Composite<T> child) => this.children.Remove(child);

    /// <summary>Depth-first, self first.</summary>
    public IEnumerable<Composite<T>> Walk()
    {
        yield return this;
        foreach (var child in this.children)
        {
            foreach (var node in child.Walk())
            {
                yield return node;
            }
        }
    }

    public double Sum(Func<T, double> of) => this.Walk().Sum(node => of(node.Value));

    public int Size => this.Walk().Count();
}

/// <summary>
/// Flyweight — share one instance per key instead of re-creating equal objects.
/// <para>
/// <c>ConcurrentDictionary.GetOrAdd</c> is what this uses; the helper adds
/// <see cref="Count"/>, <see cref="Clear"/> and a name that says what the dictionary
/// is for.
/// </para>
/// </summary>
public sealed class Flyweight<TKey, TValue>(Func<TKey, TValue> factory)
    where TKey : notnull
{
    private readonly System.Collections.Concurrent.ConcurrentDictionary<TKey, TValue> cache = new();

    public TValue this[TKey key] => this.cache.GetOrAdd(key, factory);

    public int Count => this.cache.Count;

    public void Clear() => this.cache.Clear();
}

/// <summary>
/// Bridge — a stable abstraction whose implementation can be swapped underneath it.
/// <para>
/// Callers hold the bridge and read <see cref="Api"/> per call, so a
/// <see cref="Swap"/> reaches everyone who already has the reference. Constructor
/// injection — the textbook version — makes every holder re-wire instead.
/// </para>
/// </summary>
public sealed class Bridge<TImpl, TApi>(Func<TImpl, TApi> build, TImpl implementation)
{
    public TApi Api { get; private set; } = build(implementation);

    public void Swap(TImpl next) => this.Api = build(next);
}

/// <summary>
/// Decorator — wrap a function in layers (retry, cache, log) without touching it.
/// The first wrapper is the outermost, so the arguments read in the order they run.
/// </summary>
public static class Decorator
{
    public static Func<TIn, TOut> Decorate<TIn, TOut>(
        Func<TIn, TOut> fn,
        params Func<Func<TIn, TOut>, Func<TIn, TOut>>[] wrappers)
    {
        var wrapped = fn;
        for (var i = wrappers.Length - 1; i >= 0; i--)
        {
            wrapped = wrappers[i](wrapped);
        }

        return wrapped;
    }
}

/// <summary>
/// Adapter — expose an incompatible object through the interface you want.
/// <para>
/// Thin on purpose: in C# the idiomatic adapter is a small class implementing your
/// interface, and you should usually write that. This exists so the catalog reads the
/// same across languages, and to name the intent at a call site where a lambda is enough.
/// </para>
/// </summary>
public static class Adapter
{
    public static TApi Adapt<TSource, TApi>(TSource source, Func<TSource, TApi> build) => build(source);
}

// Facade has no helper in C#, and the first attempt at one proved why: a helper that
// takes a "parts" object has to construct it to pass it in, which defeats the point.
// The language already does this properly — a class whose subsystems are Lazy<T>
// fields, each built on first read:
//
//     sealed class Checkout
//     {
//         private readonly Lazy<PaymentClient> payments = new(() => new PaymentClient(key));
//         private readonly Lazy<Mailer> mail = new(() => new Mailer(smtp));
//
//         public string Pay(int cents) => this.payments.Value.Charge(cents);
//         public string Receipt(string to) => this.mail.Value.Send(to);
//     }
//
// Nothing a helper could add. See the README table.

File: packages/kotlin/src/main/kotlin/io/github/andrewtellez/gof/Structural.kt
package io.github.andrewtellez.gof

import java.util.concurrent.ConcurrentHashMap

/**
 * Decorator — wrap a function in layers (retry, cache, log) without touching it.
 * The first wrapper is the outermost, so the list reads in the order it runs.
 *
 * ```
 * val charge = decorate(rawCharge, ::withLog, withRetry(3))
 * ```
 */
public fun <F> decorate(fn: F, vararg wrappers: (F) -> F): F =
    wrappers.foldRight(fn) { wrap, next -> wrap(next) }

/**
 * Composite — treat a tree of nodes like a single node.
 *
 * ```
 * val box = Composite(Item(price = 10.0), listOf(Composite(Item(price = 5.0))))
 * box.sum { it.price } // 15.0
 * ```
 */
public class Composite<T>(public val value: T, children: List<Composite<T>> = emptyList()) {
    private val mutableChildren: MutableList<Composite<T>> = children.toMutableList()

    public val children: List<Composite<T>> get() = mutableChildren

    public fun add(vararg nodes: Composite<T>): Composite<T> {
        mutableChildren += nodes
        return this
    }

    public fun remove(child: Composite<T>): Boolean = mutableChildren.remove(child)

    /**
     * Depth-first, self first. A [Sequence], so `filter`, `first` and early exit all
     * work without walking the rest of the tree.
     */
    public fun walk(): Sequence<Composite<T>> = sequence {
        yield(this@Composite)
        for (child in mutableChildren) yieldAll(child.walk())
    }

    public fun sum(of: (T) -> Double): Double = walk().sumOf { of(it.value) }

    public val size: Int get() = walk().count()
}

/**
 * Flyweight — share one instance per key instead of re-creating equal objects.
 *
 * ```
 * val treeTypes = Flyweight<String, TreeType> { name -> TreeType(name) }
 * treeTypes["oak"] === treeTypes["oak"] // true
 * ```
 *
 * Backed by a [ConcurrentHashMap], so it is safe to share across threads and the
 * factory runs once per key even under contention.
 */
public class Flyweight<K : Any, V : Any>(private val factory: (K) -> V) {
    private val cache: ConcurrentHashMap<K, V> = ConcurrentHashMap()

    public operator fun get(key: K): V = cache.computeIfAbsent(key) { factory(it) }

    public val size: Int get() = cache.size

    public fun clear(): Unit = cache.clear()
}
